### ISSUE TITLE
Third NameNode got added to hdfs-site.xml after adding a new head node without NamenNode role

### ISSUE THREAD

**Issue by amithkanand** [bloomberg/chef-bach#775]
I was building a cluster and I tried adding 3rd head node and two worker nodes. After successfully converging the 3rd head node I tried converging existing head nodes and for some reason the existing head node now sees 3 NameNode instead of 2 NameNodes. I see following as run_list for the 3rd head node which is different from role currently assigned to it in `cluster.txt`.

Cluster.txt has
```
bcpc-vm6 08:00:27:2C:2A:45 10.0.100.16 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Head]
```

The node object has:

```
--> knife node show bcpc-vm6.bcpc.example.com
Node Name:   bcpc-vm6.bcpc.example.com
Environment: Test-Laptop
FQDN:        bcpc-vm6.bcpc.example.com
IP:          192.168.100.16
Run List:    role[BCPC-Hadoop-Head], role[BCPC-Hadoop-Head-Namenode], role[BCPC-Hadoop-Head-HBase], recipe[bcpc-hadoop::copylog]
Roles:       BCPC-Hadoop-Head, Basic
```

The incorrect run_list seems to be causing NN failure on existing head nodes.

**Comment by http-418** [comment]
I cannot recreate this bug.  I added a fourth head node to my toy hardware cluster.

0. edit cluster.txt
1. cobbler, re-enable netboot
2. Reinstall OS
3. C-A-R basic
4. C-A-R hadoop

The run list assigned by C-A-R is correct on the node object:
` "run_list": [ "role[BCPC-Hadoop-Head]" ]`

**Comment by amithkanand** [comment]
The issue is reproducible.
```
--> cat cluster.txt 
test-laptop 08:00:27:56:A2:21 10.0.100.5 - bcpc_host bcpc.example.com SKIP
bcpc-vm1 08:00:27:56:A2:28 10.0.100.11 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Head],role[BCPC-Hadoop-Head-Namenode],role[BCPC-Hadoop-Head-HBase],recipe[bcpc-hadoop::copylog]
bcpc-vm2 08:00:27:E5:3A:00 10.0.100.12 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Head],role[BCPC-Hadoop-Head-ResourceManager],role[BCPC-Hadoop-Head-Namenode-Standby],role[BCPC-Hadoop-Head-MapReduce],role[BCPC-Hadoop-Head-Hive],recipe[bcpc-hadoop::copylog]
bcpc-vm3 08:00:27:AD:1D:EA 10.0.100.13 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Worker],recipe[bcpc-hadoop::copylog]
bcpc-vm4 08:00:27:48:E3:6D 10.0.100.14 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Worker]
bcpc-vm5 08:00:27:EE:42:72 10.0.100.15 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Worker]
bcpc-vm6 08:00:27:B9:5A:1B 10.0.100.16 - bcpc_host_trusty bcpc.example.com role[BCPC-Hadoop-Head]
[vagrant@bcpc-bootstrap (~/chef-bcpc)]
--> ./cluster-assign-roles.sh Test-Laptop hadoop bcpc-vm6





--> knife node show bcpc-vm6.bcpc.example.com
Node Name:   bcpc-vm6.bcpc.example.com
Environment: Test-Laptop
FQDN:        bcpc-vm6.bcpc.example.com
IP:          192.168.100.16
Run List:    role[BCPC-Hadoop-Head]
Roles:       BCPC-Hadoop-Head, Basic
Recipes:     bach_common::motd, bcpc::graphite_handler, ubuntu, ntp, chef-client::delete_validation, chef-client::config, bcpc::chef_vault_install, bcpc::default, bcpc::ubuntu_tools_repo, bcpc-hadoop::default, bcpc::networking, bcpc::mysql, bcpc::keepalived, bcpc::haproxy, bcpc::powerdns, bcpc-hadoop::disks, bcpc-hadoop::graphite_disk_wrapper, bcpc::graphite, bcpc::diamond, bcpc::zabbix-head, bcpc-hadoop::hdp_repo, bach_krb5::krb5_client, bcpc-hadoop::configs, pam::default, bcpc-hadoop::zookeeper_server, bcpc-hadoop::journalnode, bcpc-hadoop::graphite_to_zabbix, bcpc_jmxtrans, graphite_handler::default, chef_handler::default, ubuntu::default, apt::default, ntp::default, ntp::apparmor, logrotate::default, bcpc::cluster_local_repository, bcpc::certs, bcpc::ssh, bcpc::mysql_client, bcpc::mysql_data_bags, bfd::default, pdns::authoritative_package, build-essential::default, build-essential::_debian, pdns::backend_clients, pdns::authoritative_config, bcpc::apache2, bcpc::zabbix-repo, bcpc::zabbix-work, krb5::default, sysctl::default, sysctl::service, java::default, java::set_attributes_from_version, java::oracle, java::notify, java::set_java_home, java::default_java_symlink, java::oracle_jce, bcpc-hadoop::zookeeper_config, bcpc-hadoop::zookeeper_impl, bach_krb5::keytab_directory, bcpc-hadoop::hadoop_config, bcpc-hadoop::core_site, bcpc-hadoop::hdfs_site, bcpc-hadoop::mapred_site, bcpc-hadoop::yarn_config, bcpc-hadoop::yarn_env, bcpc-hadoop::yarn_site, bcpc-hadoop::graphite_queries, bcpc_jmxtrans::default, jmxtrans::default, ark::default
Platform:    ubuntu 14.04
Tags:        


[vagrant@bcpc-bootstrap (~/chef-bcpc)]
--> ./cluster-assign-roles.sh Test-Laptop hadoop bcpc-vm1
Recipe: bcpc-hadoop::namenode_master
  * service[hadoop-hdfs-zkfc] action restart
    - restart service service[hadoop-hdfs-zkfc]
  * service[generally run hadoop-hdfs-namenode] action restart

    ================================================================================
    Error executing action `restart` on resource 'service[generally run hadoop-hdfs-namenode]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of /etc/init.d/hadoop-hdfs-namenode restart ----
    STDOUT: * Stopping Hadoop namenode:
    stopping namenode
     * Starting Hadoop namenode:
    starting namenode, logging to /var/log/hadoop-hdfs/hadoop-hdfs-namenode-bcpc-vm1.out
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=128m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=128m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=128m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=128m; support was removed in 8.0
    Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
    STDERR:
    ---- End output of /etc/init.d/hadoop-hdfs-namenode restart ----
    Ran /etc/init.d/hadoop-hdfs-namenode restart returned 1

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/bcpc-hadoop/recipes/namenode_master.rb

    177: service "generally run hadoop-hdfs-namenode" do
    178:   action [:enable, :start]
    179:   supports :status => true, :restart => true, :reload => false
    180:   service_name "hadoop-hdfs-namenode"
    181:   subscribes :restart, "log[jdk-version-changed]", :delayed
    182:   subscribes :restart, "link[/etc/init.d/hadoop-hdfs-namenode]", :immediate
    183:   subscribes :restart, "template[/etc/hadoop/conf/hadoop-metrics2.properties]", :delayed
    184:   subscribes :restart, "template[/etc/hadoop/conf/hdfs-site.xml]", :delayed
    185:   subscribes :restart, "template[/etc/hadoop/conf/core-site.xml]", :delayed
    186:   subscribes :restart, "template[/etc/hadoop/conf/hdfs-policy.xml]", :delayed
    187:   subscribes :restart, "template[/etc/hadoop/conf/hdfs-site_HA.xml]", :delayed
    188:   subscribes :restart, "file[/etc/hadoop/conf/ldap-conn-pass.txt]", :delayed
    189:   subscribes :restart, "template[/etc/hadoop/conf/hadoop-env.sh]", :delayed
    190:   subscribes :restart, "template[/etc/hadoop/conf/topology]", :delayed
    191:   subscribes :restart, "user_ulimit[hdfs]", :delayed
    192:   subscribes :restart, "bash[initialize-shared-edits]", :immediately
    193:   subscribes :restart, "bash[hdp-select hadoop-hdfs-namenode]", :delayed
    194: end
    195:

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/bcpc-hadoop/recipes/namenode_master.rb:177:in `from_file'

    service("generally run hadoop-hdfs-namenode") do
      action [:enable, :start]
      supports {:status=>true, :restart=>true, :reload=>false}
      retries 0
      retry_delay 2
      guard_interpreter :default
      service_name "hadoop-hdfs-namenode"
      enabled true 
      running true 
      pattern "generally run hadoop-hdfs-namenode"
      cookbook_name "bcpc-hadoop"
      recipe_name "namenode_master"
    end


[vagrant@bcpc-bootstrap (~/chef-bcpc)]
--> ./nodessh.sh Test-Laptop 10.0.100.11 -

ubuntu@bcpc-vm1:~$ tail -40 /var/log/hadoop-hdfs/hadoop-hdfs-namenode-bcpc-vm1.log 
2017-02-17 09:34:48,706 INFO org.apache.hadoop.util.GSet: VM type       = 64-bit
2017-02-17 09:34:48,710 INFO org.apache.hadoop.util.GSet: 2.0% max memory 3.8 GB = 78.6 MB
2017-02-17 09:34:48,711 INFO org.apache.hadoop.util.GSet: capacity      = 2^23 = 8388608 entries
2017-02-17 09:34:48,736 INFO org.apache.hadoop.hdfs.server.blockmanagement.BlockManager: dfs.block.access.token.enable=true
2017-02-17 09:34:48,736 INFO org.apache.hadoop.hdfs.server.blockmanagement.BlockManager: dfs.block.access.key.update.interval=600 min(s), dfs.block.access.token.lifetime=600 min(s), dfs.encrypt.data.transfer.algorithm=null
2017-02-17 09:34:48,738 ERROR org.apache.hadoop.hdfs.server.namenode.FSNamesystem: FSNamesystem initialization failed.
java.lang.IllegalArgumentException: Expected exactly 2 NameNodes in namespace 'Test-Laptop'. Instead, got only 3 (NN ids were 'namenode661959116','namenode657118901','namenode666462747'
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:115)
        at org.apache.hadoop.hdfs.HAUtil.getNameNodeIdOfOtherNode(HAUtil.java:160)
        at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.createBlockTokenSecretManager(BlockManager.java:421)
        at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.<init>(BlockManager.java:316)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.<init>(FSNamesystem.java:744)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.loadFromDisk(FSNamesystem.java:680)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.loadNamesystem(NameNode.java:662)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.initialize(NameNode.java:726)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:951)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:935)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.createNameNode(NameNode.java:1641)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.main(NameNode.java:1707)
2017-02-17 09:34:48,743 INFO org.apache.hadoop.hdfs.server.namenode.FSNamesystem: Stopping services started for active state
2017-02-17 09:34:48,743 INFO org.apache.hadoop.hdfs.server.namenode.FSNamesystem: Stopping services started for standby state
2017-02-17 09:34:48,743 ERROR org.apache.hadoop.hdfs.server.namenode.NameNode: Failed to start namenode.
java.lang.IllegalArgumentException: Expected exactly 2 NameNodes in namespace 'Test-Laptop'. Instead, got only 3 (NN ids were 'namenode661959116','namenode657118901','namenode666462747'
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:115)
        at org.apache.hadoop.hdfs.HAUtil.getNameNodeIdOfOtherNode(HAUtil.java:160)
        at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.createBlockTokenSecretManager(BlockManager.java:421)
        at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.<init>(BlockManager.java:316)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.<init>(FSNamesystem.java:744)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.loadFromDisk(FSNamesystem.java:680)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.loadNamesystem(NameNode.java:662)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.initialize(NameNode.java:726)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:951)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.<init>(NameNode.java:935)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.createNameNode(NameNode.java:1641)
        at org.apache.hadoop.hdfs.server.namenode.NameNode.main(NameNode.java:1707)
2017-02-17 09:34:48,745 INFO org.apache.hadoop.util.ExitUtil: Exiting with status 1
2017-02-17 09:34:48,749 INFO org.apache.hadoop.hdfs.server.namenode.NameNode: SHUTDOWN_MSG: 
/************************************************************
SHUTDOWN_MSG: Shutting down NameNode at bcpc-vm1.bcpc.example.com/10.0.100.11
************************************************************/
ubuntu@bcpc-vm1:~$ exit
logout
[vagrant@bcpc-bootstrap (~/chef-bcpc)]
--> knife node show bcpc-vm6.bcpc.example.com
Node Name:   bcpc-vm6.bcpc.example.com
Environment: Test-Laptop
FQDN:        bcpc-vm6.bcpc.example.com
IP:          192.168.100.16
Run List:    role[BCPC-Hadoop-Head], role[BCPC-Hadoop-Head-Namenode], role[BCPC-Hadoop-Head-HBase], recipe[bcpc-hadoop::copylog]
Roles:       BCPC-Hadoop-Head, Basic
Recipes:     bach_common::motd, bcpc::graphite_handler, ubuntu, ntp, chef-client::delete_validation, chef-client::config, bcpc::chef_vault_install, bcpc::default, bcpc::ubuntu_tools_repo, bcpc-hadoop::default, bcpc::networking, bcpc::mysql, bcpc::keepalived, bcpc::haproxy, bcpc::powerdns, bcpc-hadoop::disks, bcpc-hadoop::graphite_disk_wrapper, bcpc::graphite, bcpc::diamond, bcpc::zabbix-head, bcpc-hadoop::hdp_repo, bach_krb5::krb5_client, bcpc-hadoop::configs, pam::default, bcpc-hadoop::zookeeper_server, bcpc-hadoop::journalnode, bcpc-hadoop::graphite_to_zabbix, bcpc_jmxtrans, graphite_handler::default, chef_handler::default, ubuntu::default, apt::default, ntp::default, ntp::apparmor, logrotate::default, bcpc::cluster_local_repository, bcpc::certs, bcpc::ssh, bcpc::mysql_client, bcpc::mysql_data_bags, bfd::default, pdns::authoritative_package, build-essential::default, build-essential::_debian, pdns::backend_clients, pdns::authoritative_config, bcpc::apache2, bcpc::zabbix-repo, bcpc::zabbix-work, krb5::default, sysctl::default, sysctl::service, java::default, java::set_attributes_from_version, java::oracle, java::notify, java::set_java_home, java::default_java_symlink, java::oracle_jce, bcpc-hadoop::zookeeper_config, bcpc-hadoop::zookeeper_impl, bach_krb5::keytab_directory, bcpc-hadoop::hadoop_config, bcpc-hadoop::core_site, bcpc-hadoop::hdfs_site, bcpc-hadoop::mapred_site, bcpc-hadoop::yarn_config, bcpc-hadoop::yarn_env, bcpc-hadoop::yarn_site, bcpc-hadoop::graphite_queries, bcpc_jmxtrans::default, jmxtrans::default, ark::default
Platform:    ubuntu 14.04
Tags:        
[vagrant@bcpc-bootstrap (~/chef-bcpc)]
--> 
```

**Comment by http-418** [comment]
Fixed in #782
